File: java/TripleDes.java


import javax.crypto.Cipher;
import javax.crypto.SecretKey;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import org.apache.commons.codec.binary.Base64;
public class TripleDes {
    public static byte[] TripleDesEncrypt(byte[] content, byte[] key) throws Exception {
        byte[] icv = new byte[8];
        System.arraycopy(key, 0, icv, 0, 8);
        return TripleDesEncrypt(content, key, icv);
    }
    protected static byte[] TripleDesEncrypt(byte[] content, byte[] key, byte[] icv) throws
            Exception {
        final SecretKey secretKey = new SecretKeySpec(key, "DESede");
        final Cipher cipher = Cipher.getInstance("DESede/CBC/PKCS5Padding");
        final IvParameterSpec iv = new IvParameterSpec(icv);
        cipher.init(Cipher.ENCRYPT_MODE, secretKey, iv);
        return cipher.doFinal(content);
    }
    public static byte[] TripleDesDecrypt(byte[] content, byte[] key) throws Exception {
        byte[] icv = new byte[8];
        System.arraycopy(key, 0, icv, 0, 8);
        return TripleDesDecrypt(content, key, icv);
    }
    protected static byte[] TripleDesDecrypt(byte[] content, byte[] key, byte[] icv) throws
            Exception {
        final SecretKey secretKey = new SecretKeySpec(key, "DESede");
        final Cipher cipher = Cipher.getInstance("DESede/CBC/PKCS5Padding");

        final IvParameterSpec iv = new IvParameterSpec(icv); cipher.init(Cipher.DECRYPT_MODE, secretKey, iv);
        return cipher.doFinal(content);
    }
    public static void main(String[] args) throws Exception {
        byte[] content = "{\"order_id\":\"201609010016562012987\",\"dealer_id\":\"2736123\",\"broker_id\":\"yiyun\",\"real_name\":\"张三\",\"card_no\":\"6228880199872220\",\"phone_no\":\"13488795491\",\"id_card\":\"123532612312312321\",\"pay\":\"100000.00\"}".getBytes("utf-8");
        byte[] key = "123456788765432112345678".getBytes("utf-8");
        byte[] enc = EncodeTest.TripleDesEncrypt(content, key);
        byte[] enc64 = Base64.encodeBase64(enc);
        System.out.println("encrypt: " + new String(enc64));
        byte[] dec64 = Base64.decodeBase64("0eUw2Nk2isVYjjJQE4xb65kk4HSUGy3Bp0t8PhCVy2xK6mbFfRzMt4/b0T10ZQ0dIQqY/mFQbPunq0Cy/Dqr8d3ongbX5zT7E3Rmi2PoYB2wDMjt3El+3HTtcjEAQCIJTO2budgJltbWKZ4e34GwGvWPHQgxD7bYrcSUAlUUv3S9WqVqScFa8TCJ0+WVGK3C5thx43LoU7i0YdyWZ+VsJGh3g4dsCZWprqeiLLam56bK2i0/gxqHqzVJVnAW2+WvWkJCfKBZGHPvGV/ktv5OMZYe8FVmRjZt");
        byte[] dec = EncodeTest.TripleDesDecrypt(dec64, key);
        System.out.println("decrypt: " + new String(dec));
    }
}

File: go/TripleDes.go
package main

import (
	"bytes"
	"crypto/cipher"
	"crypto/des"
	"encoding/base64"
	"encoding/json"
	"fmt"
)

func main() {
	encodeToDecode()
}

// 解密
func ThriDESDeCrypt(crypted,key []byte)[]byte{
	//获取block块
	block,_ :=des.NewTripleDESCipher(key)
	//创建切片
	context := make([]byte,len(crypted))
	//设置解密方式
	blockMode := cipher.NewCBCDecrypter(block,key[:8])
	//解密密文到数组
	blockMode.CryptBlocks(context,crypted)
	//去补码
	context = PKCSUnPadding(context)
	return context
}

//加密
func ThriDESEnCrypt(origData,key []byte)[]byte{
	//获取block块
	block,_ :=des.NewTripleDESCipher(key)
	//补码
	origData = PKCSPadding(origData, block.BlockSize())
	//设置加密方式为 3DES  使用3条56位的密钥对数据进行三次加密
	blockMode := cipher.NewCBCEncrypter(block,key[:8])
	//创建明文长度的数组
	crypted := make([]byte,len(origData))
	//加密明文
	blockMode.CryptBlocks(crypted,origData)
	//base64.StdEncoding.EncodeToString()
	return []byte(base64.StdEncoding.EncodeToString(crypted))
}
//去补码
func PKCSUnPadding(origData []byte)[]byte{
	length := len(origData)
	unpadding := int(origData[length-1])
	return origData[:length-unpadding]
}

//补码
func PKCSPadding(origData []byte,blockSize int)[]byte{
	//计算需要补几位数
	padding := blockSize-len(origData)%blockSize
	//在切片后面追加char数量的byte(char)
	padtext := bytes.Repeat([]byte{byte(padding)},padding)
	return append(origData,padtext...)
}

func encodeToDecode() {
    // 必须大于8位
	key := "123456788765432112345678"
	data := `{"order_id": "201609010016562012987","dealer_id": "2736123","broker_id":
     "yiyun","real_name": "张三","card_no": "6228880199872220","phone_no":
     "13488795491","id_card": "123532612312312321","pay": "100000.00"}`
	encrypt := ThriDESEnCrypt([]byte(data), []byte(key))
	encode, _ :=base64.StdEncoding.DecodeString(string(encrypt))
	decrypt := ThriDESDeCrypt([]byte(encode), []byte(key))
	fmt.Println(string(encrypt))
	fmt.Println(string(decrypt))
}
